### ISSUE TITLE
Puppeteer in docker container: Chromium revision is not downloaded

### ISSUE THREAD

**Issue by ebbishop** [puppeteer/puppeteer#1761]
I'm trying to launch puppeteer in an express app that's run in a docker container, using docker-compose.

### Steps to reproduce

**Tell us about your environment:**

* Puppeteer version: 0.13.0
* Platform / OS version: jessie
* Node.js version: 8.9.4

**What steps will reproduce the problem?**

### Express app's Dockerfile:

    FROM node:8

    RUN apt-get update

    # for https
    RUN apt-get install -yyq ca-certificates
    # install libraries
    RUN apt-get install -yyq libappindicator1 libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libnss3 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6
    # tools
    RUN apt-get install -yyq gconf-service lsb-release wget xdg-utils
    # and fonts
    RUN apt-get install -yyq fonts-liberation

    RUN mkdir -p /usr/src/app
    WORKDIR /usr/src/app
    COPY code/package.json /usr/src/app
    COPY code/index.js /usr/src/app
    RUN mkdir -p /usr/src/app/views
    COPY code/views/ /usr/src/app

    # install the necessary packages
    RUN yarn install

    CMD npm run start:dev

### docker-compose.yml:

    app:
        restart: always
        build: ${REPO}
        volumes:
            - ${REPO}/code:/usr/src/app:ro
        working_dir: /usr/src/app
        ports:
            - "8087:5000"

### index.js route:

    app.post('/img', function (req, res) {
      const puppeteer = require('puppeteer');
      (async() => {
        const browser = await puppeteer.launch({args: ['--no-sandbox']});
      })();
    });

**What is the expected result?**
Puppeteer can launch a browser in the express app.

**What happens instead?**

The line that should launch puppeteer `const browser = await puppeteer.launch({args: ['--no-sandbox']});` throws the following error:

`(node:28) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): AssertionError [ERR_ASSERTION]: Chromium revision is not downloaded. Run "npm install"`

I've tried adding a `yarn add puppeteer` after the `yarn install` in the Dockerfile. I've also tried replacing `yarn install` in the `Dockerfile` with `npm install` .  Neither makes a difference.

**Comment by ebidel** [comment]
Assuming your package.json has `puppeteer` has a dependency.... Can you `docker exec` into the container and verify if there's a populated `./node_modules/puppeteer/.local-chromium/` folder?

**Comment by ebbishop** [comment]
There is:
`./node_modules/puppeteer/.localchromium/mac-508693/chrome-mac/Chromium.app/Contents`
- Info.plist
- MacOS
- PkgInfo
- Resources
- Versions

**Comment by ebidel** [comment]
That path looks like it's from a mac.  If you're in Docker, it should be something like `./node_modules/puppeteer/.local-chromium/linux-515411/`.

**Comment by ebbishop** [comment]
That's it. The last line in `volumes` in the `docker-compose.yml` was throwing it off by loading everything from my machine to the docker machine.

**Comment by Lazzza** [comment]
Did you find a solution for this I am experiencing the same issue, what did you have to change in your `docker-compose.yml `

**Comment by kempsterrrr** [comment]
I'm also having this same problem, if you could post your solution that would be awesome.

If I figure this out I will post it up here too

**Comment by yllieth** [comment]
Entering the container, chrome is here ... but not there is the same time:
```bash
pwd
/opt/us-app/node_modules/puppeteer/.local-chromium/linux-594312/chrome-linux

ls -l /opt/us-app/node_modules/puppeteer/.local-chromium/linux-594312/chrome-linux
total 259148
drwxr-sr-x    2 us-app   ubiservi      4096 Oct 26 15:49 MEIPreload
-rwxr-xr-x    1 us-app   ubiservi 215709936 Oct 26 15:49 chrome
-rwxr-xr-x    1 us-app   ubiservi      5087 Oct 26 15:49 chrome-wrapper
-rw-r--r--    1 us-app   ubiservi   1015884 Oct 26 15:49 chrome_100_percent.pak
-rw-r--r--    1 us-app   ubiservi   1336011 Oct 26 15:49 chrome_200_percent.pak
-rwxr-xr-x    1 us-app   ubiservi    318840 Oct 26 15:49 chrome_sandbox
-rw-r--r--    1 us-app   ubiservi  10246144 Oct 26 15:49 icudtl.dat
-rwxr-xr-x    1 us-app   ubiservi    353368 Oct 26 15:49 libEGL.so
-rwxr-xr-x    1 us-app   ubiservi   7207056 Oct 26 15:49 libGLESv2.so
-rwxr-xr-x    1 us-app   ubiservi   3512520 Oct 26 15:49 libclearkeycdm.so
drwxr-sr-x    2 us-app   ubiservi      4096 Oct 26 15:49 locales
-rwxr-xr-x    1 us-app   ubiservi   3807744 Oct 26 15:49 nacl_helper
-rwxr-xr-x    1 us-app   ubiservi      9608 Oct 26 15:49 nacl_helper_bootstrap
-rwxr-xr-x    1 us-app   ubiservi   3838116 Oct 26 15:49 nacl_helper_nonsfi
-rwxr-xr-x    1 us-app   ubiservi   3868208 Oct 26 15:49 nacl_irt_x86_64.nexe
-rw-r--r--    1 us-app   ubiservi     92225 Oct 26 15:49 natives_blob.bin
-rw-r--r--    1 us-app   ubiservi      2540 Oct 26 15:49 product_logo_48.png
drwxr-sr-x    3 us-app   ubiservi      4096 Oct 26 15:49 resources
-rw-r--r--    1 us-app   ubiservi  13166555 Oct 26 15:49 resources.pak
drwxr-sr-x    2 us-app   ubiservi      4096 Oct 26 15:49 swiftshader
-rw-r--r--    1 us-app   ubiservi    745272 Oct 26 15:49 v8_context_snapshot.bin
-rwxr-xr-x    1 us-app   ubiservi     37394 Oct 26 15:49 xdg-mime
-rwxr-xr-x    1 us-app   ubiservi     33273 Oct 26 15:49 xdg-settings

./chrome
bash: line 4: ./chrome: No such file or directory
```

**Comment by sinahwz** [comment]
> That path looks like it's from a mac. If you're in Docker, it should be something like `./node_modules/puppeteer/.local-chromium/linux-515411/`.

I am on a windows machine, any ideas how I can fix the issue? 
The folder in my case is -> 'win64-579032'

**Comment by sinahwz** [comment]
The manual solution is to `docker exec` into the container, delete the puppeteer folder from node_modules. and 'npm install' again. now I have 'linux-579032'.
But still need a solution to fix it without the manual hassle.

**Comment by dylanjha** [comment]
Running into this on mac, here's my solution for now:

Getting this error inside the docker container after running `docker compose up`:

> Chromium revision is not downloaded.

```shell
# find the docker process id
docker ps

// start a new bash session inside the docker container
docker exec -it <process-id> bash

# view contents of puppeteer ./node_modules/puppeteer/.local-chromium
# I can see mac-624492 is installed

ls ./node_modules/puppeteer/.local-chromium

# re-install dependencies
yarn install

# view contents of puppeteer ./node_modules/puppeteer/.local-chromium
# I can now that linux-624492 is installed
# no longer getting "Chromium revision is not downloaded" error

ls ./node_modules/puppeteer/.local-chromium
```

**Comment by omidkrad** [comment]
If you are connecting to an instance of puppeteer in a linked container, you should use `puppeteer.connect()` instead of `puppeteer.launch()`, or you may get that "Chromium revision is not downloaded." error.

**Comment by catalinberta** [comment]
My quick fix was to manually run `install.js` from the installed npm puppeteer package:
Go inside the container: `docker run --rm -it -u root:root -w $PWD -v $PWD:$PWD directory-name /bin/bash`
Then go inside where puppeteer is installed in `node_modules` and run the install script `node install.js` - for some reason it made it work and no longer throw `Chromium revision is not downloaded` even though it was already downloaded

**Comment by arturoromeroslc** [comment]
https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md#running-puppeteer-in-docker

**Comment by esraajb** [comment]
I experience that if I add volume to the puppeteer service in docker compose

`volumes:`
 `     - "./e2e:/e2e"`

**Comment by wallacepreston** [comment]
I don't know where I found this solution (it was another GitHub issue).  I am running node 12 alpine inside a docker container, and these 3 lines were necessary right before `RUN npm install`.  This (1) Installs chromium so it's available in the container, (2) tells puppeteer not to install chromium, and (3) tells puppeteer what path it can find chromium at.
```
RUN apk add -U --no-cache --allow-untrusted udev ttf-freefont chromium git
ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD true
ENV CHROMIUM_PATH /usr/bin/chromium-browser
```
Then when I use puppeteer in my JS code, I have to tell it what path to use (the ENV var I set above).  
```
const puppeteer = require('puppeteer');

const browser = await puppeteer.launch({
	headless: true,
	executablePath: process.env.CHROMIUM_PATH,
	args: ['--no-sandbox'],
});
```
Note: I didn't hardcode the path, since I want to use a different path per OS (I don't always run my app in Docker)
